05.ls/ls.rb

```diff
@@ -28,24 +29,103 @@ def display_columns(columns)
 end
 
 def ls(params)
-  dirs_and_files = Dir.glob('*')
+  files = Dir.glob('*', sort: true)
 
-  dirs_and_files.sort!
-  dirs_and_files.reverse! if params[:r]
-
-  unless (dirs_and_files.size % NUM_OF_ROWS).zero?
-    (NUM_OF_ROWS - dirs_and_files.size % NUM_OF_ROWS).times do
-      dirs_and_files << ''
+  unless (files.size % NUM_OF_ROWS).zero?
+    (NUM_OF_ROWS - files.size % NUM_OF_ROWS).times do
+      files << ''
     end
   end
 
   columns = []
-  dirs_and_files.each_slice(dirs_and_files.size / NUM_OF_ROWS) do |column|
-    columns << column
+  unless files.empty?
+    files.each_slice(files.size / NUM_OF_ROWS) do |column|
+      columns << column
+    end
   end
 
   adjust_digits(columns)
-  display_columns(columns)
+  params[:l] ? ls_long_format : display_columns(columns)
+end
+
+def change_filetype_notation(file_with_stat)
+  case file_with_stat[0][0, 3]
+  when '010'
+    file_with_stat[0][0, 3] = 'p'
+  when '020'
+    file_with_stat[0][0, 3] = 'c'
+  when '040'
+    file_with_stat[0][0, 3] = 'd'
+  when '060'
+    file_with_stat[0][0, 3] = 'b'
+  when '100'
+    file_with_stat[0][0, 3] = '-'
+  when '120'
+    file_with_stat[0][0, 3] = 'l'
+  when '140'
+    file_with_stat[0][0, 3] = 's'
+  end
+end
+
+def change_mode_notation(file_with_stat)
+  file_with_stat[0][1..3].each_char do |char|
+    case char
+    when '0'
+      file_with_stat[0].gsub!(/0/, '---')
+    when '1'
+      file_with_stat[0].gsub!(/1/, '--x')
+    when '2'
+      file_with_stat[0].gsub!(/2/, '-w-')
+    when '3'
+      file_with_stat[0].gsub!(/3/, '-wx')
+    when '4'
+      file_with_stat[0].gsub!(/4/, 'r--')
+    when '5'
+      file_with_stat[0].gsub!(/5/, 'r-x')
+    when '6'
+      file_with_stat[0].gsub!(/6/, 'rw-')
+    when '7'
+      file_with_stat[0].gsub!(/7/, 'rwx')
+    end
+  end
+end
+
+def find_max_digit(stat_attribute)
+  stat_attribute.max.to_s.length
+end
+
+def display_files(files_with_stat, file_nlinks, file_sizes)
+  files_with_stat.each do |file_with_stat|
+    change_filetype_notation(file_with_stat)
+    change_mode_notation(file_with_stat)
```

Comment:
破壊的な処理をやめて「引数に渡した変数を更新する」をやめれば、ここはメソッドが返す値を変数に格納する形になると思います。こんなイメージでしょうか。
```suggestion
    filetypes = change_filetype_notation(file_with_stat)
    modes = change_mode_notation(file_with_stat)
```